### PR TITLE
reduce torch.zeros overhead in moe align block size kernel

### PR DESCRIPTION
**Pull request by BBuf**
## Motivation

- [x] reduce torch.zeros overhead in moe align block size kernel.
- [x]  There is a small improvement when bs<=16, for example, when bs=range(1, 15), the `fused_moe` kernel speed improvement magnitude is 0.3/0.28=1%.

origin:

图片

pr:

图片

## Benchmark

main:

```shell
fused-moe-performance:
     batch_size  vllm_fused_moe_triton  sglang_fused_moe_triton
0           1.0               0.355232                 0.306560
1           2.0               0.345664                 0.303840
2           3.0               0.343264                 0.299968
3           4.0               0.345696                 0.304192
4           5.0               0.347824                 0.305152
5           6.0               0.347136                 0.304768
6           7.0               0.347840                 0.308256
7           8.0               0.349984                 0.308576
8           9.0               0.349504                 0.308288
9          10.0               0.350080                 0.308576
10         11.0               0.350560                 0.308112
11         12.0               0.355456                 0.307680
12         13.0               0.366944                 0.309232
13         14.0               0.378208                 0.306624
14         15.0               0.381824                 0.309808
15         16.0               0.469888                 0.306464
16         17.0               0.475616                 0.305440
```

pr:

```shell
     batch_size  vllm_fused_moe_triton  sglang_fused_moe_triton
0           1.0               0.344144                 0.283072
1           2.0               0.340368                 0.280368
2           3.0               0.338864                 0.277792
3           4.0               0.342304                 0.283440
4           5.0               0.347936                 0.284928
5           6.0               0.340288                 0.281808
6           7.0               0.341248                 0.282720
7           8.0               0.347392                 0.284640
8           9.0               0.342144                 0.284608
9          10.0               0.350176                 0.287008
10         11.0               0.349024                 0.284832
11         12.0               0.348352                 0.286816
12         13.0               0.366480                 0.285312
13         14.0               0.378784                 0.284448
14         15.0               0.380800                 0.286848
15         16.0               0.468640                 0.290736
16         17.0               0.474432                 0.302128
```

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py [hunk 1]
```diff
@@ -650,6 +650,54 @@ def moe_align_block_size_triton(
     )
 
 
+@triton.jit
+def init_sorted_ids_and_cumsum_buffer_kernel(
+    sorted_ids_ptr,
+    cumsum_buffer_ptr,
+    max_num_tokens_padded: tl.constexpr,
+    topk_ids_numel: tl.constexpr,
```
Comment (Alcanderian) [inline python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: I wonder whether using max_num_tokens_padded and topk_ids_numel as constexpr will create too many kernels without yielding significant performance benefits

Comment (BBuf) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: The shape of `topk_ids` is `[num_tokens, topk]`. Under CUDA graph mode, num_tokens is fixed, and `num_experts/topk` is also fixed, so there should be no risk of kernel recompilation.

Comment (Alcanderian) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: > The shape of `topk_ids` is `[num_tokens, topk]`. Under CUDA graph mode, num_tokens is fixed, and `num_experts/topk` is also fixed, so there should be no risk of kernel recompilation.

But in prefill stage num_tokens is very variant...

Comment (BBuf) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: Right, if we need to consider prefill, does that mean we can only resolve it with a CUDA kernel?

Comment (Alcanderian) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: Will it be very slow without constexpr? 

And suggestions, BLOCK_SIZE = 1024
```
sorted_ids_blocks = tl.num_programs(0) - 1
```
```
elif pid == sorted_ids_blocks:
    offset_e = tl.arange(0, ALIGNED_NUM_EXPERTS_P1)
    mask_e = offset_e < num_experts + 1
    tl.store(cumsum_buffer_ptr + offset_e, tl.zeros((ALIGNED_NUM_EXPERTS_P1,), dtype=tl.int32)
```

Comment (BBuf) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: Make sense, I can set `ALIGNED_NUM_EXPERTS_P1` to 512?

Comment (BBuf) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: Testing shows removing those two constexpr has negligible performance impact. While the absolute values fluctuate in the data below, the relative ratios remain nearly equal between both versions.

before:

图片

after:

图片

Comment (Alcanderian) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: > Make sense, I can set `ALIGNED_NUM_EXPERTS_P1` to 512?

triton.next_power_of_2(num_experts + 1)

Comment (BBuf) [reply python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py]: done!

**Review by Alcanderian — APPROVED** [review]

**Review by merrymercy — APPROVED** [review]
